### PR TITLE
Avoid precondition failure in write timeout

### PR DESCRIPTION
**Pull request by rnro**
### Motivation:

In some cases we can crash because of a precondition failure when the write timeout fires and we aren't in the running state. This can happen for example if the connection is closed whilst the write timer is active.

### Modifications:

* Remove the precondition and instead take no action if the timeout fires outside of the running state. Instead we take a new `Action`, `.noAction` when the timer fires.
* Clear write timeouts upon request completion. When a request completes we have no use for the idle write timer, we clear the read timer and we should clear the write one too.

### Result:

Fewer crashes.

The supplied tests fails without these changes and passes with either of them.

### CODE REVIEW

File: Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ClientChannelHandler.swift [hunk 1]
```diff
@@ -281,7 +281,7 @@ final class HTTP1ClientChannelHandler: ChannelDuplexHandler {
         case .close:
             context.close(promise: nil)
 
-        case .wait:
+        case .wait, .noAction:
```
Comment (rnro) [inline Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ClientChannelHandler.swift:284]: Do we need this new state or can we just re-use `.wait`?

Comment (fabianfett) [reply Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ClientChannelHandler.swift:284]: reuse `.wait`. `.wait` is a do nothing. can be renamed in a follow up pr.

File: Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ConnectionStateMachine.swift [hunk 2]
```diff
@@ -359,7 +361,7 @@ struct HTTP1ConnectionStateMachine {
 
     mutating func idleWriteTimeoutTriggered() -> Action {
         guard case .inRequest(var requestStateMachine, let close) = self.state else {
-            preconditionFailure("Invalid state: \(self.state)")
+            return .noAction
```
Comment (fabianfett) [inline Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ConnectionStateMachine.swift:364]: Do you mind adding another test to HTTP1ConnectionStateMachineTests that just tests this edge?

File: Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ConnectionStateMachine.swift [hunk 3]
```diff
@@ -83,6 +83,8 @@ struct HTTP1ConnectionStateMachine {
         case fireChannelActive
         case fireChannelInactive
         case fireChannelError(Error, closeConnection: Bool)
+
+        case noAction
```
Comment (fabianfett) [inline Sources/AsyncHTTPClient/ConnectionPool/HTTP1/HTTP1ConnectionStateMachine.swift:87]: please remove.

File: Tests/AsyncHTTPClientTests/HTTP1ClientChannelHandlerTests.swift [hunk 4]
```diff
@@ -840,6 +840,81 @@ class HTTP1ClientChannelHandlerTests: XCTestCase {
         channel.writeAndFlush(request, promise: nil)
         XCTAssertEqual(request.events.map(\.kind), [.willExecuteRequest, .requestHeadSent])
     }
+
+    class SlowHandler: ChannelOutboundHandler {
```
Comment (fabianfett) [inline Tests/AsyncHTTPClientTests/HTTP1ClientChannelHandlerTests.swift:844]: this isn't used at all.

File: Tests/AsyncHTTPClientTests/HTTP1ClientChannelHandlerTests.swift [hunk 5]
```diff
@@ -840,6 +840,81 @@ class HTTP1ClientChannelHandlerTests: XCTestCase {
...
+
+        // start a request stream we'll never write to
+        let streamPromise = embedded.eventLoop.makePromise(of: Void.self)
+        let streamCallback = { @Sendable (streamWriter: HTTPClient.Body.StreamWriter) -> EventLoopFuture<Void> in
+            streamPromise.futureResult
+        }
+        request.body = .init(contentLength: nil, stream: streamCallback)
+
+        let delegate = NullResponseDelegate()
+        var maybeRequestBag: RequestBag<NullResponseDelegate>?
```
Comment (fabianfett) [inline Tests/AsyncHTTPClientTests/HTTP1ClientChannelHandlerTests.swift:874]: Why not use a `ResponseAccumulator` here instead?

**Review by fabianfett — COMMENTED** [review]
Nits.

File: Tests/AsyncHTTPClientTests/HTTP1ClientChannelHandlerTests.swift (hunk 5 above)
Comment (rnro) [reply Tests/AsyncHTTPClientTests/HTTP1ClientChannelHandlerTests.swift]: I hadn't spotted that, thanks.
